crypto/odocrypt.c

```diff
@@ -0,0 +1,300 @@
+//
+//  odocrypt.c
+//  DigiByte
+//
+//  Created by Julian Jäger on 22.07.19.
+//  Copyright © 2019 DigiByte Foundation NZ Limited. All rights reserved.
+//
+
+#include "odocrypt.h"
+#include "sha3/BRKeccak-800-SnP.h"
+
+#include <assert.h>
+#include <string.h>
+
+uint64_t Rot(uint64_t x, int r)
+{
+    return r == 0 ? x : (x << r) ^ (x >> (64-r));
+}
+
+void OdoCrypt_Unpack(uint64_t* state, const char* bytes)
+{
+    memset(state, 0, STATE_SIZE * sizeof(uint64_t));
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        for (int j = 0; j < 8; j++)
+        {
+            state[i] |= (uint64_t)(uint8_t)bytes[8*i + j] << (8*j);
+        }
+    }
+}
+
+void OdoCrypt_PreMix(uint64_t* state)
+{
+    uint64_t total = 0;
+    for (int i = 0; i < STATE_SIZE; i++)
+        total ^= state[i];
+    total ^= total >> 32;
+    for (int i = 0; i < STATE_SIZE; i++)
+        state[i] ^= total;
+}
+
+void OdoCrypt_ApplyMaskedSwaps(uint64_t* state, const uint64_t* mask)
+{
+    for (int i = 0; i < STATE_SIZE/2; i++)
+    {
+        uint64_t a = state[2*i];
+        uint64_t b = state[2*i+1];
+        // For each bit set in the mask, swap the corresponding bits in `a` and `b`
+        uint64_t swp = mask[i] & (a ^ b);
+        a ^= swp;
+        b ^= swp;
+        
+        state[2*i] = a;
+        state[2*i+1] = b;
+    }
+}
+
+void OdoCrypt_ApplyWordShuffle(uint64_t* state, int m)
+{
+    uint64_t next[STATE_SIZE];
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        next[m*i % STATE_SIZE] = state[i];
+    }
+    memcpy(state, &next[0], STATE_SIZE * sizeof(uint64_t));
+}
+
+void OdoCrypt_ApplyPboxRotations(uint64_t* state, const int* rotation)
+{
+    for (int i = 0; i < STATE_SIZE/2; i++)
+    {
+        // Only rotate the even words.  Rotating the odd words wouldn't actually
+        // be useful - a transformation that rotates all the words can be
+        // transformed into one that only rotates the even words, then rotates
+        // the odd words once after the final iteration.
+        state[2*i] = Rot(state[2*i], rotation[i]);
+    }
+}
+
+void OdoCrypt_ApplyPbox(uint64_t* state, const Pbox* perm)
+{
+    for (int i = 0; i < PBOX_SUBROUNDS-1; i++)
+    {
+        // Conditionally move bits between adjacent pairs of words
+        OdoCrypt_ApplyMaskedSwaps(state, perm->mask[i]);
+        // Move the words around
+        OdoCrypt_ApplyWordShuffle(state, PBOX_M);
+        // Rotate the bits within words
+        OdoCrypt_ApplyPboxRotations(state, perm->rotation[i]);
+    }
+    OdoCrypt_ApplyMaskedSwaps(state, perm->mask[PBOX_SUBROUNDS-1]);
+}
+
+void OdoCrypt_Pack(const uint64_t* state, char* bytes)
+{
+    memset(bytes, 0, ODOCRYPT_DIGEST_SIZE * sizeof(char));
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        for (int j = 0; j < 8; j++)
+        {
+            bytes[8*i + j] = (state[i] >> (8*j)) & 0xff;
+        }
+    }
+}
+
+void OdoCrypt_ApplySboxes(uint64_t* state, const uint8_t** sbox1, const uint16_t** sbox2)
```

Comment:
I think there's a problem trying to cast a `uint8_t[SMALL_SBOX_COUNT][1 << SMALL_SBOX_WIDTH]` to `const uint8_t**`.

Reply:
Oh! That sounds true. Would you like to help me with that one?